File: repositories/otp_repository.go
package repositories

import (
	"context"
	"errors"
	"time"

	"doctor-service/entities"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type OTPRepository interface {
	Create(ctx context.Context, otp *entities.OTPVerification) error
	FindLatestActive(ctx context.Context, doctorID uuid.UUID, purpose entities.OTPPurpose) (*entities.OTPVerification, error)
	IncrementAttempt(ctx context.Context, otpID uuid.UUID) error
	MarkVerified(ctx context.Context, otpID uuid.UUID) error
	InvalidatePreviousOTPs(ctx context.Context, doctorID uuid.UUID, purpose entities.OTPPurpose) error
}

type GormOTPRepository struct {
	db *gorm.DB
}

func NewOTPRepository(db *gorm.DB) OTPRepository {
	return &GormOTPRepository{db: db}
}

func (r *GormOTPRepository) Create(ctx context.Context, otp *entities.OTPVerification) error {
	return r.db.WithContext(ctx).Create(otp).Error
}

func (r *GormOTPRepository) FindLatestActive(ctx context.Context, doctorID uuid.UUID, purpose entities.OTPPurpose) (*entities.OTPVerification, error) {
	var otp entities.OTPVerification
	err := r.db.WithContext(ctx).
		Where("doctor_id = ? AND purpose = ? AND verified_at IS NULL", doctorID, purpose).
		Order("created_at DESC").
		First(&otp).Error

	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, nil
	}
	return &otp, err
}

func (r *GormOTPRepository) IncrementAttempt(ctx context.Context, otpID uuid.UUID) error {
	return r.db.WithContext(ctx).
		Model(&entities.OTPVerification{}).
		Where("id = ?", otpID).
		UpdateColumn("attempt_count", gorm.Expr("attempt_count + 1")).Error
}

func (r *GormOTPRepository) MarkVerified(ctx context.Context, otpID uuid.UUID) error {
	now := time.Now()
	return r.db.WithContext(ctx).
		Model(&entities.OTPVerification{}).
		Where("id = ?", otpID).
		Update("verified_at", &now).Error
}

func (r *GormOTPRepository) InvalidatePreviousOTPs(ctx context.Context, doctorID uuid.UUID, purpose entities.OTPPurpose) error {
	now := time.Now()
	return r.db.WithContext(ctx).
		Model(&entities.OTPVerification{}).
		Where("doctor_id = ? AND purpose = ? AND verified_at IS NULL", doctorID, purpose).
		Update("expires_at", now).Error
}

File: python_backend/repositories/otp_repository.py
import datetime
from typing import Optional
from uuid import UUID
from sqlalchemy.orm import Session
from entities.otp import OTPVerification

class OTPRepository:
    def __init__(self, db: Session):
        self.db = db

    def Create(self, otp: OTPVerification) -> None:
        self.db.add(otp)
        self.db.commit()
        self.db.refresh(otp)

    def FindLatestActive(self, doctor_id: UUID, purpose: str) -> Optional[OTPVerification]:
        return self.db.query(OTPVerification).filter(
            OTPVerification.doctor_id == str(doctor_id),
            OTPVerification.purpose == purpose,
            OTPVerification.is_verified == False
        ).order_by(OTPVerification.created_at.desc()).first()

    def IncrementAttempt(self, otp_id: UUID) -> None:
        self.db.query(OTPVerification).filter(OTPVerification.id == str(otp_id)).update({
            "attempt_count": OTPVerification.attempt_count + 1
        })
        self.db.commit()

    def MarkVerified(self, otp_id: UUID) -> None:
        self.db.query(OTPVerification).filter(OTPVerification.id == str(otp_id)).update({"is_verified": True})
        self.db.commit()

    def InvalidatePreviousOTPs(self, doctor_id: UUID, purpose: str) -> None:
        now = datetime.datetime.now()
        self.db.query(OTPVerification).filter(
            OTPVerification.doctor_id == str(doctor_id),
            OTPVerification.purpose == purpose,
            OTPVerification.is_verified == False
        ).update({"expires_at": now})
        self.db.commit()
